cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java

```diff
@@ -203,61 +210,144 @@ public void initialize() throws Exception {
+          });
+
+      } else if (BatchJoiner.PLUGIN_TYPE.equals(pluginType)) {
+
+        final BatchJoiner<?, ?, ?> batchJoiner = pluginInstantiator.newPluginInstance(stageName, evaluator);
+        ContextProvider<DefaultJoinerContext> contextProvider =
+          new JoinerContextProvider(pipelineRuntime, stageSpec, context.getAdmin());
+        submitterPlugin = new SubmitterPlugin<>(
+          stageName, context, batchJoiner, contextProvider,
+          new SubmitterPlugin.PrepareAction<DefaultJoinerContext>() {
+            @Override
+            public void act(DefaultJoinerContext joinerContext) {
+              if (joinerContext.getNumPartitions() != null) {
+                job.setNumReduceTasks(joinerContext.getNumPartitions());
+              }
+              Class<?> outputKeyClass = joinerContext.getJoinKeyClass();
+              Class<?> inputRecordClass = joinerContext.getJoinInputRecordClass();
+
+              if (outputKeyClass == null) {
+                outputKeyClass = TypeChecker.getJoinKeyClass(batchJoiner);
+              }
+              if (inputRecordClass == null) {
+                inputRecordClass = TypeChecker.getJoinInputRecordClass(batchJoiner);
+              }
+              hConf.set(MAP_KEY_CLASS, outputKeyClass.getName());
+              hConf.set(MAP_VAL_CLASS, inputRecordClass.getName());
+              job.setMapOutputKeyClass(getOutputKeyClass(stageName, outputKeyClass));
+              getOutputValClass(stageName, inputRecordClass);
+              // for joiner plugin map output is tagged with stageName
+              job.setMapOutputValueClass(TaggedWritable.class);
+            }
+          });
       }
-      try {
-        BatchConfigurable<BatchSinkContext> batchSink = pluginInstantiator.newPluginInstance(sinkName, evaluator);
-        MapReduceBatchContext sinkContext = new MapReduceBatchContext(context, pipelineRuntime, stageInfo,
-                                                                      connectorDatasets);
-        batchSink.prepareRun(sinkContext);
-        finishers.add(batchSink, sinkContext);
-        sinkOutputs.put(sinkName, new SinkOutput(sinkContext.getOutputNames(), stageInfo.getErrorDatasetName()));
-      } catch (Exception e) {
-        // Catch the Exception to generate a User Error Log for the Pipeline
-        PIPELINE_LOG.error("Failed to initialize batch sink '{}' with the error: {}. Please review your pipeline " +
-                            "configuration and check the system logs for more details.", sinkName,
-                          Throwables.getRootCause(e).getMessage(), Throwables.getRootCause(e));
-        throw e;
+      if (submitterPlugin != null) {
+        submitterPlugin.prepareRun();
+        finishers.add(submitterPlugin);
       }
     }
-    finisher = finishers.build();
+
     hConf.set(SINK_OUTPUTS_KEY, GSON.toJson(sinkOutputs));
+    hConf.set(INPUT_ALIAS_KEY, GSON.toJson(inputAliasToStage));
+    finisher = finishers.build();
```

Comment:
+1 for moving all the `finishers.add` to before the call to `finishers.builder`, even though calling the methods in the reverse order was also working...